README
OpenCaster 

==Contact==
Lorenzo Pallara (l.pallara@avalpa.com)
Andrea Venturi (a.venturi@avalpa.com)

==Introduction==
This package is a collection of softwares for Linux system to play out and multiplex MPEG2 transport stream format
Use-cases:
MHP, MHEG5, HbbTv and others interactive tv standards object carousel broadcast
Playout of locally stored, offline encoded audio and video for non-live tv and radio stations
Multiplexing of input multicasted UDP transport stream to output multicasted UDP transport stream
VOD system based on mpeg2 transport stream over IP or over DVB for walled garden network like hotels with coax or ethernet.
DVB-SSU update
Data casting
Teletext generator
IP encapsualtor
PSI/SI generator
EPG generator

==Integration==
OpenCaster output has been successfully integrated with a long list of dvb products, among them there are products by Cisco/Scientific Atlanta, Ericsson/Tandberg, Dektec, Eurotek, Harmonic/Scopus, MainConcept, Mitan, Screen Service, Adtec, Wellav, Arris, Sr-Systems, Deltacast, ...

==Manual==
Avalpa Broadcast Server manual is available here: http://www.avalpa.com/the-key-values/15-free-software/33-opencaster

==Test==
OpenCaster has been tested with Rohde & Schwarz DVM100L, Rohde & Schwarz DVM 400, Dektec StreamXpert and a countless number of set top boxes, digital receivers and television sets

==Technical features==
- PSI generation (PAT, PMT, NIT, SDT, TDT, AIT, TOT, ...)
- EPG/EIT generation
- Object and data carousel generation and receivement (MHP, MHEG5, HbbTv, DVB-SSU, DSMCC, ...)
- third party software AC-3, MPEG2 audio, MPEG2 video and H264 video encoders support
- Audio Description generation support
- IP encapsulation (MPE)
- Teletext generation
- CBR transport stream muxing
- PCR stamping
- Null packet replacement
- PID filtering
- PID remapping
- Null packet insertion
- Video buffer verifier
- PCR timing verifier
- PES headers analyzer
- MPEG2 video and MPEG2/AC-3 audio ES header analyzer
- M2TS file generation from TS file

==Demo==
A demonstration transport stream is available at this address:http://www.avalpa.com/the-key-values/15-free-software/59-opencaster-demo-roll

==Handy tables==

DVB-T modulations:
Modulation 	Coding rate 	Guard interval
				1/4 	1/8 	1/16 	1/32
QPSK 		1/2 		4.976 	5.529 	5.855 	6.032
		2/3 		6.635 	7.373 	7.806 	8.043
		3/4 		7.465 	8.294 	8.782 	9.048
		5/6 		8.294 	9.216 	9.758 	10.053
		7/8 		8.709 	9.676 	10.246 	10.556
16-QAM 		1/2 		9.953 	11.059 	11.709 	12.064
		2/3 		13.271 	14.745 	15.612 	16.086
		3/4 		14.929 	16.588 	17.564 	18.096
		5/6 		16.588 	18.431 	19.516 	20.107
		7/8 		17.418 	19.353 	20.491 	21.112
64-QAM 		1/2 		14.929 	16.588 	17.564 	18.096
		2/3 		19.906 	22.118 	23.419 	24.128
		3/4 		22.394 	24.882 	26.346 	27.144
		5/6 		24.882 	27.647 	29.273 	30.160
		7/8 		26.126 	29.029 	30.737 	31.668

channel to frequency table:
21 	474
22 	482
23 	490
24 	498
25 	506
26 	514
27 	522
28 	530
29 	538
30 	546
31 	554
32 	562
33 	570
34 	578
35 	586
36 	594
37 	602
38 	610
39 	618
40 	626
41 	634
42 	642
43 	650
44 	658
45 	666
46 	674
47 	682
48 	690
49 	698
50 	706
51 	714
52 	722
53 	730
54 	738
55 	746
56 	754
57 	762
58 	770
59 	778
60 	786
61 	794
62 	802
63 	810
64 	818
65 	826
66 	834
67 	842
68 	850
69 	858

mpeg audio layer 2 suggested bit rates:
56 kbps -> 1 pes packet -> 1 ts packet
112 kbps -> 1 pes packet -> 2 ts packets 
160 kbps -> 1 pes packet -> 3 ts packets
224 kbps -> 1 pes packet -> 4 ts packets
all the others fall in the middle and are not a wise choice


code/CHANGES
21/09/2013

Change log author:
Lorenzo Pallara (l.pallara@avalpa.com)

Change log OpenCaster 3.2.1:
- tools/*: m2ts2ts renamed m2ts2cbrts
- tools/tsprinter: removed
- libs/dvbobjects/*: added support for ATSC EIT
- tutorials/*: removed tutorials from main packet, they are avilable on a support download from http://www.avalpa.com
- extra/*: removed, they are now part of support download from http://www.avalpa.com

Change log OpenCaster 3.2:
- tools/mpe2sec/: added suppor for 64bit (by Julian Cable julian.cable@bbc.co.uk)
- libs/dvbobjects/*: added more dvb-ssu descriptors support (by Andreas Regel andreas.regel@ttgoerler.de)
- libs/dvbobjects/*: added BAT support (by Andreas Regel andreas.regel@ttgoerler.de)
- libs/dvbobjects/*: removed duplicated descriptor ISO_639_language_descriptor
- libs/dvbobjects/*: added application_recording_descriptor (by Jesper Gustin jesperg@opera.com)
- tools/txt2pes.c: fixed compatibility issue 
- tools/tsdiscont: added to fix some discontiuity issues (by Shane Andrusiak, blendz@shaw.ca)
- tools/tsstamp: added pusi check

Change log OpenCaster 3.1.4:
- tutorials/*: capitalized ISO639 language codes 
- tutorials/hbbtv-http: fixed small typo
- tutorials/firsttime: bug fix for localtime
- libs/dvbobjects/dvbobjects/DSMCC/Message.py compatibility fix for mythtv

Change log OpenCaster 3.1.3:
- tools/pesvideo2ts: source code cosmetics
- extras/x264: added patch for x264 to output format required by OpenCaster
- tutorials/hbbtv-dsmcc: added descriptor required by some receiveres

Change log OpenCaster 3.1.2
- tutorial/eitfilter: bug fix for services without scheduled events
- tools/oc-update.sh: improved carousel head cache chances on some decoders
- libs/: allows application_signaling with empty subfields

Change log OpenCaster 3.1.1:
- tools/*: fix overflow on CLOCK_MONOTONIC functions

Change log OpenCaster 3.1:
- tutorials: added .srt to teletext subtitles example, documentation coming soon
- libs/: bug fix for multiplex txt pages in pmt descriptor
- tools/esaudio2pes: added support for DTS, still on test
- tools/esaudioinfo: added support for DTS, still on test
- libs: added DVB descriptors for DTS
- libs: bug fixes to support python 2.6
- tools/* replaced gettimeofday with clock_gettime to enforce CLOCK_MONOTIC
- tools/eitsecactualtoanother: added
- tools/tstdt: added start time option for replay 
- tools/tstcpsend: bug fix on init. value
- tutorials: removed some useless file
- libs: added SIT table
- tutorials: added SIT example

Change log OpenCaster 3.0:
- tools/pesvideo2ts: generalized vbv, command line supports vbv size in bit, now it broadcasts h264 video (join is on test)
- tools/pesvideo2ts: added 24 fps support
- tools/ts2m2ts: added generate m2ts from ts
- tools/m2ts2ts: added generate ts from m2ts
- tools/pesinfo: prints also video pes size
- tools/esvideo2pes renamed tools/esvideompeg2pes
- tools/esvideoinfo renamed tools/esvideompeg2info
- tools/vbv renamed tools/videompeg2vbv
- tutorials: renamed with a meaningful name

Change log OpenCaster 2.5.1:
- libs/dvbobjects: removed default value for segment_last_section_number in EIT generation
- tools/tspcrrestamp: rewritten, multicast multiplexing should recover better in network downs
- tools/tsccc: doesn't check anymore for null packets cc errors 
- tutorials/OCtutorial14: TOT example manages DST with iCalendar RFC
- tutorials/OCtutorial12/13: removed as not supported anymore

Change log OpenCaster 2.5:
- tutorials/[mhp]: now include new Avalpa's Wizard xlet
- tools/oc-update.sh: introduced new "smart compress" option, improved carousel optimization, check details inside
- tools/file2mod.y: introduced new "smart compress" option, improved carousel optimization, check details inside

Change log OpenCaster 2.4.7
- tools/eitsecfilter: fix deprecated include for <limits.h>

Change log OpenCaster 2.4.6.1:
- tutorials/: some bug fixes

Change log OpenCaster 2.4.6:
- libs/dvbobjects/HBBTV: Added HbbTv support by Claude Vanderm (claude.vanderm@mediatv.com)
- tools/eitsecfilter: save eit section on different file names ready for remuxing and playout
- tools/eitsecmapper: replace onid and tsid on eit section files
- tools/tsprinter: outputs a transport stream over a pc parallel port pins (max 2.1mbps 1 cpu 1GHZ 100%, ts packet jitter is high, use only for data)
- tools/tsmodder: bug fix on ts input loop
- tools/dvbobjects: bug fix for ATSC descriptors and tutorial
- tools/ts2sec: added crc32 check, discard invalid sections
- tools/pesaudio2ts: support fps other than 25 fps
- tools/pesvideo2ts: support fps other than 25 fps
- tools/pesclock: discontinued
- tools/oc2sec: help changed, module version was not reported as hex
- tools/tsstamp: header bug fix by (blendz@shaw.ca)
- tools/tsmask: command line bug fix by (kailife.forge@gmail.com)
- tutorials/: added tutorials for HbbTV
- tutorials/: added tutorial for epg filtering and muxing
- libs/dvbobjecs: added content descriptor
- libs/dvbobjecs: bug fix registration_descriptor
- libs/dvbobjecs/DVB/DataCarousel: bug fix for more than 256 modules by Robert Antosiak (robert.antosiak@digitaltv-labs.com)

Change log OpenCaster 2.4.5:
- tools/esvideo2pes: bug fix, now support fps other than 25 fps

Change log OpenCaster 2.4.4:
- tools/pesvideo2ts: added option to change pcr repetition rate
- libs/dvbobjects: transaction id update flag bug fix (by tdv team at lifia.info.unlp.edu.ar)
- libs/dvbobjects: added descriptors for SDBT (by tdv team at lifia.info.unlp.edu.ar)
- tutorials:  data change, some TS7900HD firmware versions doesn't seem to support remount of the same carousel

Change log OpenCaster 2.4.3:
- tools/pesaudio2ts: bug fix on exit without loop, support for audio description generation
- tools/esaudio2pes: support for audio description generation
- tools/dvbobjects: bug fix for language descriptor, added support for audio description generation
- tools/esvideoinfo: detects end_sequence
- tools/pesdata2ts: bug fix ( was not using write)

code/tools/mpe2sec/mpe.c
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <syslog.h>
#include <errno.h>
#include <arpa/inet.h>

#include <sys/ioctl.h>
#include <sys/socket.h>
#include <net/if.h>

#include <net/if_tun.h>

#include "sectioncrc.h"

/* pre 2.4.6 compatibility */
#define OTUNSETNOCSUM  (('T'<< 8) | 200)
#define OTUNSETDEBUG   (('T'<< 8) | 201)
#define OTUNSETIFF     (('T'<< 8) | 202)
#define OTUNSETPERSIST (('T'<< 8) | 203)
#define OTUNSETOWNER   (('T'<< 8) | 204)

static int persist = 0;
static int stuff = 0;
static char padding[184];
static char ip_device[IFNAMSIZ];
static char s[180];
static const char *Id = "$Id: mpe.c 25 2011-10-13 15:35:18Z jfbcable $";
const MPE_HEADER_LEN=12;
int tun_fd = -1;

#ifdef IFF_TUN
int tun_open(char *dev)
{
    struct ifreq ifr;
    int fd;

    if ((fd = open("/dev/net/tun", O_RDWR)) < 0)
    {
        perror("open tun");
        return -1;
    }

    memset(&ifr, 0, sizeof(ifr));
    ifr.ifr_flags = IFF_TUN;
    if(persist == 0)
      ifr.ifr_flags |= IFF_NO_PI;
    if (*dev)
        strncpy(ifr.ifr_name, dev, IFNAMSIZ);
    else
    {
        fprintf(stderr, "device name not supplied\n");
        return -1;
    }

    if (ioctl(fd, TUNSETIFF, (void *) &ifr) < 0) {
        if (errno == EBADFD) {
            /* Try old ioctl */
            if (ioctl(fd, OTUNSETIFF, (void *) &ifr) < 0)
            {
                perror("new nor old ioctl worked\n");
                close(fd);
                return -1;
            }
        } else
	{
                perror("other error\n");
                close(fd);
                return -1;
	}
    }

    strncpy(dev, ifr.ifr_name, IFNAMSIZ);
    return fd;

}
#else
int tun_open(char *dev)
{
    char tunname[14];
    int i, fd, err;

    if( *dev ) {
       sprintf(tunname, "/dev/%s", dev);
       return open(tunname, O_RDWR);
    }

    sprintf(tunname, "/dev/tun");
    err = 0;
    for(i=0; i < 255; i++){
       sprintf(tunname + 8, "%d", i);
       /* Open device */
       if( (fd=open(tunname, O_RDWR)) > 0 ) {
          strcpy(dev, tunname + 5);
          return fd;
       }
       else if (errno != ENOENT)
          err = errno;
       else if (i)      /* don't try all 256 devices */
          break;
    }
    if (err)
        errno = err;
    perror("no /dev/tun* available\n");
    return -1;
}

#endif

void send_mpe(int fd, unsigned char *buf, size_t len)
{
    unsigned char *mpe_header = buf;
    unsigned char *ip_datagram = &buf[MPE_HEADER_LEN];
    unsigned long crc;
    unsigned long i;
    unsigned short section_len = len - 3 + 4; // table id and section len not in, crc in!
    mpe_header[0] = 0x3e;
    mpe_header[1] = ((section_len >> 8) & 0x0f) | 0xb0;
    mpe_header[2] = section_len & 0xff;
    mpe_header[3] = 0;
    mpe_header[4] = 0;
    mpe_header[5] = 0xc1;
    mpe_header[6] = 0;
    mpe_header[7] = 0;
    mpe_header[8] = 0;
    mpe_header[9] = 0;
    mpe_header[10] = 0;
    mpe_header[11] = 0;
    if ((ip_datagram[16] & 0xe0) == 0xe0) {	/* multicast */
        mpe_header[3] = ip_datagram[19];
        mpe_header[4] = ip_datagram[18];
        mpe_header[8] = ip_datagram[17] & 0x7f;
        mpe_header[9] = 0x5e;
        mpe_header[10] = 0;
        mpe_header[11] = 1;
    }
    crc = htonl(sectioncrc(buf, len));
    memcpy(&buf[len], &crc, 4);
    len += 4;
    write(fd, buf, len);
    if (stuff) {
        unsigned long stuff_count = 184 - (len % 184) - 1; /* one less for TS packet pointer */
        if (stuff_count > 0) {
            write(fd, padding, stuff_count);
        }
    }
}

void usage(char **argv)
{
    fprintf(stderr, "usage %s [-p] [-s] devname\n", argv[0]);
    fprintf(stderr,
            "Create a tun device and send DVB/MPE DSM-CC sections to stdout.\n");
    fprintf(stderr,
            "-p don't create or configure the tun device\n");
    fprintf(stderr,
            "-s stuff sections to multiple of 184 bytes using 0xff octets\n");
    fprintf(stderr,
            "Project home page http://code.google.com/p/dvb-mpe-encode\n");
    fprintf(stderr, "Example:\nmpe dvb0 | sec2ts 430 | DtPlay 1000000\n");
    exit(1);
}


void exit_program(int sig)
{
    fprintf(stderr, "stopping %s\n", ip_device);
    sprintf(s, "ifdown %s", ip_device);
    system(s);
    close(tun_fd);
    exit(0);
}

int main(int argc, char **argv)
{
    int n = 1;
    if (argc < 2)
        usage(argv);
    if (argc > 4)
        usage(argv);
    while(argv[n][0] == '-')
    {
        if (strcmp(argv[n], "-p") == 0)
		persist = 1;
        if (strcmp(argv[n], "-s") == 0)
	{
		stuff = 1;
		memset(padding, 0xff, sizeof(padding));
	}
	n++;
    }
    strcpy(ip_device, argv[n]);
    tun_fd = tun_open (ip_device);
    if (tun_fd == -1)
        usage(argv);
    (void) signal(SIGINT, exit_program);
    (void) signal(SIGQUIT, exit_program);
    if(persist==0)
    {
      sprintf(s, "ifup %s", ip_device);
      system(s);
    }
    while (1) {
        unsigned char buf[4100];
        unsigned char *mpe_header = buf;
        unsigned char *tun_header = &buf[MPE_HEADER_LEN];
        int n = read(tun_fd, tun_header, sizeof(buf));
        //write(2, tun_header, n);
        send_mpe(1, mpe_header, n + MPE_HEADER_LEN);
    }
}
